### PR TITLE
Release v2.4.0

### PR DESCRIPTION
**Pull request by fsimonis**
## How to work with this template

* [x] assign a release manager, who takes care of the process
* [x] assign each point below to a responsible person, before you continue. Use `@member`.

Only the release manager should update this post (even tickboxes, due to race conditions in editing). Everybody else should comment on the PR with the progress.

## Pre-PR steps

* [x] Look over [PRs](https://github.com/precice/precice/pulls?q=is%3Apr+no%3Amilestone+is%3Aclosed) and [Issues](https://github.com/precice/precice/pulls?q=is%3Apr+no%3Amilestone+is%3Aclosed) without an assigned version. (all)
* [x] Look over entries in [`docs/changelog`](https://github.com/precice/precice/blob/develop/docs/changelog)) (all)
   * Add missing entries, if necessary
   * Fix wording and tense
* [x] Make sure you have the latest `develop` and `master` branches locally.
* [x] Merge master to develop ( This should result in no commits )
* [x] Check code base w.r.t code formatting (run [`precice/tools/formatting/check-format`](https://github.com/precice/precice/blob/develop/tools/formatting/check-format)) and reformat if required (run [`precice/tools/formatting/format-all`](https://github.com/precice/precice/blob/develop/tools/formatting/format-all))
* [x] Create branch `release-vX.Y.Z` from develop. If needed, `git rebase develop`.
* [x] Run `tools/releasing/bumpversion.sh MAJOR.MINOR.PATCH` to bump the version
* [x] Look over [`CHANGELOG.md`](https://github.com/precice/precice/blob/develop/CHANGELOG.md) (all)
   * Check for merged lines
   * Add things, if necessary
   * Fix wording and tense
   * Sort the entries lexicographically
   * Extract summary
* [x] Verify the version changes in:
   * [x] [CHANGELOG](https://github.com/precice/precice/blob/develop/CHANGELOG.md)
   * [x] [CMakeLists.txt](https://github.com/precice/precice/blob/develop/CMakeLists.txt)
   * [x] [debian changelog](https://github.com/precice/precice/blob/develop/tools/releasing/packaging/debian/changelog)
* [x] Commit the version bump
* [x] Push the release branch to the precice repository
* Prepare independent releases
   * [ ] [Python bindings](https://github.com/precice/python-bindings/blob/develop/docs/ReleaseGuide.md)
   * [ ] (if necessary!) [MATLAB bindings](https://github.com/precice/matlab-bindings/blob/develop/docs/ReleaseGuide.md)
   * [ ] (if necessary!) [JULIA bindings](https://github.com/precice/PreCICE.jl)

## Step by step guide

* [x] Open PR from `release-vX.Y.Z` to `master` (use [this template](https://github.com/precice/precice/blob/add_PR_template/.github/PULL_REQUEST_TEMPLATE/release_pull_request_template.md))
* [x] Do regression tests using the release branch (specific revision) _list below :arrow_down:_ (all)
* [x] Fix potential problems in develop (all)
* [x] Rebase the release branch on develop (version bump should be the latest commit)
* [ ] Draft message to mailing list
* [x] Write a draft "blog post" on [Discourse](https://precice.discourse.group/)
/blob/master/_includes/xmlreference.md)
* [x] Approve the PR with at least two reviews (all)
* [x] Merge PR to master ( use `git merge --no-ff release-vX.Y.Z` )
* [x] Tag release on master `vX.Y.Z` and verify by running `git describe --tags`
* [ ] Merge back to develop and verify by running `git describe --tags`
* [ ] Tripple check that you haven't messed anything up. (You can always discard local changes)
* [ ] Push master and push the `vX.Y.Z` tag
* [ ] Push develop
* [ ] Wait for the release pipeline
  * [ ] [To create a new draft release on GitHub](https://github.com/precice/precice/releases)
  * [ ] To automatically generate packages for latest Debian and two latest Ubuntu LTS versions.
* [ ] Write release text
* [ ] Publish the GitHub release

## Regression Tests

Use the following branches:
* precice `release-vX.Y.Z`
* pyprecice `python-bindings-vX.Y.Z.1`
* matlab-bindings `matlab-bindings-vX.Y.Z.1`
* rest `master`

Run all these tests manually on your system. If you succeed, please write a comment with the revisions of the components that you used below. Example: https://github.com/precice/precice/pull/507#issuecomment-530432289 and update the table.

| State | Success | Failure | Skipped |
| --- | --- | --- | --- |
| Write | `:o:` | `:x:` | `:fast_forward:` |
| Read | :o: | :x: | :fast_forward: |

| State | Tester | Test |
| --- | --- | --- |
| :o: :o:  | @IshaanDesai @uekerman | [perpendicular-flap](https://github.com/precice/tutorials/tree/master/perpendicular-flap) fluid-nutils - solid-calculix |
| :o: | @DavidSCN  | [perpendicular-flap](https://github.com/precice/tutorials/tree/master/perpendicular-flap) fluid-openfoam - solid-dealii |
| :o: | @IshaanDesai  | [perpendicular-flap](https://github.com/precice/tutorials/tree/master/perpendicular-flap) fluid-su2 - solid-fenics |
| :o: | @DavidSCN  | [multiple-perpendicular-flaps](https://github.com/precice/tutorials/tree/master/multiple-perpendicular-flaps) fluid-openfoam - solid-(left+right)-dealii |
| :o:  | @MakisH  | [flow-over-heated-plate](https://github.com/precice/tutorials/tree/master/flow-over-heated-plate) fluid-openfoam - solid-openfoam serial + parallel |
| :o: | @IshaanDesai  | [flow-over-heated-plate](https://github.com/precice/tutorials/tree/master/flow-over-heated-plate) fluid-openfoam - solid-fenics serial + parallel |
| :o: :o: | @IshaanDesai @uekerman   | [flow-over-heated-plate](https://github.com/precice/tutorials/tree/master/flow-over-heated-plate) fluid-openfoam - solid-nutils |
| :o: | @MakisH  | [flow-over-heated-plate-nearest-projection](https://github.com/precice/tutorials/tree/master/flow-over-heated-plate-nearest-projection) fluid-openfoam - solid-openfoam |
| :o: | @IshaanDesai   | [flow-over-heated-plate-steady-state](https://github.com/precice/tutorials/tree/master/flow-over-heated-plate-steady-state) fluid-openfoam - solid-codeaster |
| :x: :o: | @MakisH @fsimonis  | [heat-exchanger](https://github.com/precice/tutorials/tree/master/heat-exchanger) fluid-(inner+outer)-openfoam - solid-calculix |
| :o: | @MakisH  | [partitioned-elastic-beam](https://github.com/precice/tutorials/tree/master/partitioned-elastic-beam) dirichlet-calculix - neumann-calculix |
| :o: :o: | @IshaanDesai @uekerman  | [partitioned-heat-conduction](https://github.com/precice/tutorials/tree/master/partitioned-heat-conduction) fenics - nutils |
|:o: :o: | @IshaanDesai @uekerman   | [partitioned-heat-conduction-complex](https://github.com/precice/tutorials/tree/master/partitioned-heat-conduction-complex) fenics- fenics |
| :o: | @MakisH  | [partitioned-pipe](https://github.com/precice/tutorials/tree/master/partitioned-pipe) fluid1-openfoam-pimplefoam - fluid2-openfoam-sonicliquidfoam |
| :o: | @MakisH  | [elastic-tube-1d](https://github.com/precice/tutorials/tree/master/elastic-tube-1d) fluid-cpp - solid-python |
| :o: | @MakisH | [elastic-tube-3d](https://github.com/precice/tutorials/tree/master/elastic-tube-3d) fluid-openfoam - solid-calculix |
| :o: | @erikscheurer   | MATLAB / MATLAB [ODEs](https://github.com/precice/matlab-bindings/tree/develop/tutorial) |
| :o: | @MakisH | Solverdummy [Fortran module](https://github.com/precice/fortran-module/tree/develop/examples/solverdummy) | 
| :o: | @IshaanDesai  | Solverdummy [Python](https://github.com/precice/python-bindings/tree/develop/solverdummy) |
| :o:  | @erikscheurer  | Solverdummy [MATLAB](https://github.com/precice/matlab-bindings/tree/develop/solverdummy) |
| :o: | @IshaanDesai   | Solverdummy [Julia](https://github.com/precice/PreCICE.jl/tree/develop/solverdummy) |
| :o: | @uekerman  | Alya |
| :o: |  @KyleDavisSA | SuperMUC |

## Post-release

* [ ] Update documentation
    * [ ] Install from source
    * [ ] Baseline Ubuntu
    * [ ] Master-tag to intra-comm
    * [ ] Solverdummies (mesh arg)
* [ ] Flag [Arch Linux AUR package](https://aur.archlinux.org/packages/precice) and dependants as out-of-date.
* [ ] Update Spack recipe
* [ ] Update pyprecice Spack
* [ ] Update Website:
    * [ ] Bump version in [`_config.yml`](https://github.com/precice/precice.github.io/blob/master/_config.yml) and [`_data/sidebars/docs_sidebar.yml`](https://github.com/precice/precice.github.io/blob/master/_data/sidebars/docs_sidebar.yml)
    * [ ] Update the [XML reference](https://github.com/precice/precice.github.io/blob/master/_includes/xmlreference.md) using `binprecice md`
    * [ ] Look over the [Roadmap](https://www.precice.org/fundamentals-roadmap.html) and update entries.

### Release new version for bindings (to ensure compatibility with newest preCICE version)

- [ ] [Fortan module](https://github.com/precice/fortran-module/compare/master...develop) @MakisH 
- [ ] [MATLAB bindings](https://github.com/precice/matlab-bindings/blob/develop/docs/ReleaseGuide.md) @IshaanDesai 
- [ ] [python bindings](https://github.com/precice/python-bindings/blob/develop/docs/ReleaseGuide.md) @IshaanDesai 
- [ ] [Julia bindings](https://github.com/precice/PreCICE.jl) @IshaanDesai 

### Marketing

* [ ] Finalize post on [Discourse](https://precice.discourse.group/) (all)
* [ ] Write on [Gitter](https://gitter.im/precice/Lobby) @fsimonis 
* [ ] Send announcement to the [mailing list](https://mailman.informatik.uni-stuttgart.de/mailman/listinfo/precice) @fsimonis 
* [ ] [CFD-Online](https://www.cfd-online.com/Forum/news.cgi/form/0) @MakisH 
* [ ] NADigest @KyleDavisSA 
* [ ] Post on [Twitter](https://twitter.com/preCICE_org) (additionally to the automatic) @MakisH 
* [ ] Post on [ResearchGate](https://www.researchgate.net/project/preCICE) @MakisH 
* [ ] Post in [LinkedIn Group](https://www.linkedin.com/groups/9073912/) @IshaanDesai 
* [ ] Submit a short article to the [Quartl](https://www.in.tum.de/en/i05/further-activities/quartl/) DL mid june @BenjaminRodenberg 

### Misc

* [ ] Update the [PR template](https://github.com/precice/precice/blob/add_PR_template/.github/PULL_REQUEST_TEMPLATE/release_pull_request_template.md)

To open a new PR with this template, use this [PR template query](https://github.com/precice/precice/compare/new?template=release_pull_request_template.md)

### CODE REVIEW

**Comment by DavidSCN** [comment]
All my assigned test cases work (tested in serial and parallel). My setup:

- preCICE in release mode v2.4.0
- Ubuntu 22.04
- deal.II 9.4.0-pre (revision 91bff6e62a) and deal.II adapter revision 005933d
- openfoam2012 and openfoam-adapter revision [a95da52](https://github.com/precice/openfoam-adapter/commit/a95da52062c2f503ce58d9509c3e47f5efdbf149)

I cannot test HAWK any more as the project has expired.

**Comment by MakisH** [comment]
We have several warning at the moment, many of which are not from headers. Is that known and intentional? We have, for example, `-Wunused-variable`, `-Wreorder`, `-Wsign-compare`, `-Wunused-but-set-variable`. We also have a few deprecation warnings from our tests, which I guess just means that we still test deprecated features, so it is fine.

[build.log](https://github.com/precice/precice/files/8654211/build.log)

**Comment by MakisH** [comment]
> @MakisH good catch. The reorder warnings and the sign compares should be fixed. The unused warnings are pointless to fix until we get maybe_unused with C++17.

If we are bumping the default support to Ubuntu 20.04 (GCC 9.3.0), then why not directly use that? It should be supported [since GCC 7](https://en.cppreference.com/w/cpp/compiler_support/17).

**Comment by MakisH** [comment]
All tests with:
- Ubuntu 21.10, with GCC 11.2.0
- preCICE from v2.4.0 (56a50fe, based on 805fbdd), built in Debug mode
- OpenFOAM v2112 and adapter v1.1.0
- CalculiX 2.19 and adapter v2.19.0 (with `-fallow-argument-mismatch` in `FFLAGS`)
- Python bindings 2.3.0.1 (yes, it just works)
- Fortran module 38fe542 (distribution v2202.0)
- tutorials v202202.0
- running completely and qualitatively checking the results in ParaView

Results:
- OK: flow-over-heated-plate fluid-openfoam - solid-openfoam serial
- OK: flow-over-heated-plate fluid-openfoam - solid-openfoam parallel
- OK: flow-over-heated-plate-steady-state fluid-openfoam - solid-openfoam serial
- OK: flow-over-heated-plate-steady-state fluid-openfoam - solid-openfoam parallel
- BROKEN: heat-exchanger fluid-(inner+outer)-openfoam - solid-calculix
  - It fails on my system, but [this is not the first time I see this](https://github.com/precice/tutorials/issues/260). I suggest that someone else also tries it (preferably not on Ubuntu 20.04, we already have previous data for that in the issue)
- OK: partitioned-elastic-beam dirichlet-calculix - neumann-calculix
- OK: partitioned-pipe fluid1-openfoam-pimplefoam - fluid2-openfoam-sonicliquidfoam
- OK: elastic-tube-1d fluid-cpp - solid-python
- OK: elastic-tube-3d fluid-openfoam - solid-calculix
- OK: Solverdummy Fortran module

**Comment by MakisH** [comment]
> @MakisH good catch. The reorder warnings and the sign compares should be fixed. The unused warnings are pointless to fix until we get maybe_unused with C++17.

Working on it...


> > @MakisH good catch. The reorder warnings and the sign compares should be fixed. The unused warnings are pointless to fix until we get maybe_unused with C++17.
> 
> If we are bumping the default support to Ubuntu 20.04 (GCC 9.3.0), then why not directly use that? It should be supported [since GCC 7](https://en.cppreference.com/w/cpp/compiler_support/17).

...including this

**Comment by IshaanDesai** [comment]
All tutorials are run with:

* Ubuntu 20.04
* preCICE v2.4.0 release branch: https://github.com/precice/precice/commit/56a50feb38c4383865c73cbd8741c78185b4b852
* Tutorials repository: https://github.com/precice/tutorials/commit/17a8d914cbd644213d13f4f8e1c845e7d1eaea1d
* CalculiX 2.19 and adapter: https://github.com/precice/calculix-adapter/commit/069d3776d5743fe28a43cf1fef4242e20619bb3c
* OpenFOAM v2012 and adapter: https://github.com/precice/openfoam-adapter/commit/a95da52062c2f503ce58d9509c3e47f5efdbf149
* FEniCS v.2019 and adapter: https://github.com/precice/fenics-adapter/commit/8863cd9e4488da1f52f93749f320ea1a405e2a96

The following cases run properly till the end and are OK:

* perpendicular-flap: fluid-su2 and solid-fenics
* flow-over-heated-plate: fluid-openfoam and solid-fenics serial
* flow-over-heated-plate: fluid-openfoam and solid-fenics parallel

**Comment by IshaanDesai** [comment]
The Nutils scripts from `flow-over-heated-plate` and `perpendicular-flap` are not working for me with Nutils v6.0, v6.1, v6.2, v6.3. I am unsure if it is something specific to my system and hence @uekerman will also test them.

**Comment by KyleDavisSA** [comment]
Building on superMUC-NG works:

- Boost 1.73.0
- PETSc 3.13.0
- Eigen 3.3.7
- Python 3.8.8

cpp solverdummies ran successfully.

**Comment by uekerman** [comment]
I tested the Nutils cases:

* `flow-over-heated-plate`
* `perpendicular-flap`
* `partitioned-heat-conduction`

all from the `v202202.0` release.  Works with Nutils 6.3.

**Comment by uekerman** [comment]
Alya case works

**Comment by IshaanDesai** [comment]
In the interest of time I tested the solverdummy of the Julia bindings, which @erikscheurer manages. Testing with https://github.com/precice/PreCICE.jl/commit/86cbdef0a0b683579d4fc64d35e987041a436bd4 and it works :+1:

**Comment by fsimonis** [comment]
Ok, let's ignore the Matlab bindings. There haven't been major changes in the API and we tested it via Julia and python.

**Comment by erikscheurer** [comment]
Tests with:
* Ubuntu 20.04
* precice 2.4.0 release branch: [`52230261f`](https://github.com/precice/precice/commit/52230261f1ba9849ce23e4c2a41aa76e8ef95140)
* matlab bindings: [`796f138`](https://github.com/precice/matlab-bindings/commit/796f1386e77c680990a3e2236d664cd9fd54ab72)

The following cases run properly till the end and are OK:
* Solverdummy [MATLAB](https://github.com/precice/matlab-bindings/tree/develop/solverdummy)
* MATLAB / MATLAB [ODEs](https://github.com/precice/matlab-bindings/tree/develop/tutorial)

**Review by IshaanDesai — APPROVED** [review]

**Review by KyleDavisSA — APPROVED** [review]

**Review by DavidSCN — DISMISSED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,42 @@
...
+- Added experimental API functions for waveform interpolation. Currently restricted to parallel-implicit coupling. Refer to the [user documentation](https://precice.org/couple-your-code-waveform.html). (https://github.com/precice/precice/pull/1187)
+- Added export to CSV using `<export:csv />`. (https://github.com/precice/precice/pull/1144)
+- Added export to VTP using `<export:vtp />`. (https://github.com/precice/precice/pull/1137)
+- Added export to VTU using `<export:vtu />`. (https://github.com/precice/precice/pull/1136)
+- Added the `support-radius` as an additional configuration option to the Gaussian RBF mapping configuration. (https://github.com/precice/precice/pull/1163)
+- Added the API header `precice/Tooling.hpp`, which includes utility functions. (https://github.com/precice/precice/pull/1122)
+- Added the command `check` to `binprecice`, which checks a given configuration file for correctness. (https://github.com/precice/precice/pull/1132)
+- Added the command `version` to `binprecice`, which prints the version string of the used preCICE library. (https://github.com/precice/precice/pull/1122)
```
Comment (MakisH) [inline CHANGELOG.md:19]: The renaming to `precice-tools` seems to be missing.

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,42 @@
...
+- Improved the efficiency of vertex to edge projections. (https://github.com/precice/precice/pull/1226)
+- Migrated to the [FindPython3](https://cmake.org/cmake/help/v3.16/module/FindPython3.html) module for locating Python3 and NumPy. (https://github.com/precice/precice/pull/1263)
+- Migrated to using the target provided by [FindLibXml2](https://cmake.org/cmake/help/v3.16/module/FindLibXml2.html) (https://github.com/precice/precice/pull/1263)
+- Removed meshName from solverdummies input parameters. (https://github.com/precice/precice/pull/1256)
+- Renamed all instances of master to primary and of slave to secondary, along with their plural and joint forms. (https://github.com/precice/precice/pull/1253)
+- Renamed class `MasterSlave` to `IntraComm` and changed all corresponding names. Changed the XML tag `master` to `intra-comm` and changed the XML attributes `on-slaves` and `on-master` to `on-secondary-ranks` and `on-primary-rank` respectively. The tag `master` and attributes `on-slaves` and `on-master` are deprecated. (https://github.com/precice/precice/pull/1274)
```
Comment (MakisH) [inline CHANGELOG.md:38]: Should we maybe comment that some compiler warnings here are expected and nothing to worry about?

**Review by BenjaminRodenberg — APPROVED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (uekerman) [reply CHANGELOG.md:38]: IIRC those are related to `mapReadDataFrom` etc. (deprecation already older), not to these changes.

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,42 @@
...
+- Added export to CSV using `<export:csv />`. (https://github.com/precice/precice/pull/1144)
+- Added export to VTP using `<export:vtp />`. (https://github.com/precice/precice/pull/1137)
+- Added export to VTU using `<export:vtu />`. (https://github.com/precice/precice/pull/1136)
+- Added the `support-radius` as an additional configuration option to the Gaussian RBF mapping configuration. (https://github.com/precice/precice/pull/1163)
+- Added the API header `precice/Tooling.hpp`, which includes utility functions. (https://github.com/precice/precice/pull/1122)
+- Added the command `check` to `binprecice`, which checks a given configuration file for correctness. (https://github.com/precice/precice/pull/1132)
+- Added the command `version` to `binprecice`, which prints the version string of the used preCICE library. (https://github.com/precice/precice/pull/1122)
+- Added tooling to CMake, which simplifies contributing. (https://github.com/precice/precice/pull/1143)
+- Changed baseline from Ubuntu 18.04 LTS to Ubuntu 20.04 LTS. preCICE now requires Boost version `1.71.0` and CMake version `3.16.3`. (https://github.com/precice/precice/pull/1259)
```
Comment (MakisH) [inline CHANGELOG.md:21]:
```suggestion
- Renamed `binprecice` to `precice-tools`. (https://github.com/precice/precice/pull/1175)
- Changed baseline from Ubuntu 18.04 LTS to Ubuntu 20.04 LTS. preCICE now requires Boost version `1.71.0` and CMake version `3.16.3`. (https://github.com/precice/precice/pull/1259)
```

**Comment by fsimonis** [comment]
The heat-exchanger tutorial case runs in the VM with preCICE 2.4.0 installed.

File: CHANGELOG.md (hunk 2 above)
Comment (fsimonis) [reply CHANGELOG.md:38]: This was deprecated in https://github.com/precice/precice/commit/8f56303cca8206524fc4c6cd7a850926de713777, which is part of this release. So, it should be mentioned.

File: CHANGELOG.md (hunk 3 above)
Comment (fsimonis) [reply CHANGELOG.md:21]: Added an extended version.

File: CHANGELOG.md (hunk 2 above)
Comment (fsimonis) [reply CHANGELOG.md:38]: Done

File: CHANGELOG.md (hunk 1 above)
Comment (fsimonis) [reply CHANGELOG.md:19]: It's listed in the "renaming" part :smiley:

**Comment by fsimonis** [comment]
Thanks to @MakisH for pointing out some issues in the changelog. I corrected them and will need 2 more approvals.

**Review by DavidSCN — APPROVED** [review]

**Review by IshaanDesai — APPROVED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (MakisH) [reply CHANGELOG.md:19]: Not before :smirk:

**Pull request merged**